### PR TITLE
Adds the Botanitank(TM), a juiced up grinder with integrated reagent storage

### PR DESCRIPTION
**Pull request by Migratingcocofruit**
## What Does This PR Do

Adds the Botanitank. A  replacement to botany's reagent grinder with higher capacity and speed, built in reagent storage, and the ability to dispense each reagent individually to a connected beaker.  
Cannot grind minerals outside of glass and iron as those can be grown. 

## Why It's Good For The Game

This together with #25050 will lead to botany no longer needing a chem dispenser. 

## Images of changes

image

## Testing

changed between grinder circuit board types
took the Botanitank apart and put it back together
ground and juiced a bunch of foodstuffs
dispensed and dumped various reagent mixes
made sure the sprites behave as intended
ground some patches and cigarettes to make sure I didn't break that PR
tried to grind slime extracts and some minerals and couldn't, as expected. 

## Changelog
:cl:
add: Added the Botanitank
imageadd: Added icons for the Botanitank
:cl:

### CODE REVIEW

**Comment by 1080pCat** [comment]
any way you could implement a rounding down function? so you're not stuck with 46.78966454 u of banana juice

File: code/game/machinery/constructable_frame.dm
Comment (1080pCat) [inline code/game/machinery/constructable_frame.dm:675]:
```suggestion
							"Botanitank" = /obj/machinery/reagentgrinder/Botanitank)
```

**Comment by Migratingcocofruit** [comment]
> any way you could implement a rounding down function? so you're not stuck with 46.78966454 u of banana juice

You'd simply be left with less reagents this way, and you can always dump the excess. Also, this only occurs if you dispense all reagents proportionally(dispense mix button) which you wouldn't be doing without a specific purpose. 

Perhaps I'll do something for the UI.

File: code/game/machinery/constructable_frame.dm
Comment (1080pCat) [inline code/game/machinery/constructable_frame.dm:679]:
```suggestion
	if(!I.use_tool(src, user, FALSE, volume = I.tool_volume))
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (1080pCat) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:536]:
```suggestion
// The Botanitank, a bigger, faster grinder complete with integrated chem storage tank.
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (1080pCat) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:550]:
```suggestion

```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (1080pCat) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:742]:
```suggestion
	playsound(loc, 'sound/machines/blender.ogg', 50, TRUE)
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (1080pCat) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:861]:
```suggestion
	playsound(src.loc, 'sound/machines/juicer.ogg', 20, TRUE)
```

**Comment by 1080pCat** [comment]
> You'd simply be left with less reagents this way, and you can always dump the excess. Also, this only occurs if you dispense all reagents proportionally(dispense mix button) which you wouldn't be doing without a specific purpose.
> 
> Perhaps I'll do something for the UI.

sort of my want, getting stuck with .564756 can get very frustrating with doing reactions

**Comment by farie82** [comment]
Does it have a higher build cost or something? I can't find it quickly while glancing the code.

**Comment by AyIong** [comment]
I don't think it's necessary to create a separate interface for this, you can just add a Tank section to ReagentGrinder if it's a BotaniTank. Look at smartfridge/composter

**Comment by CinnamonSnowball** [comment]
1000u is a bit crazy for storage, maybe tone it down to 300u (highest we typically go, with Bluespace Beakers and fermenting barrels), or even just 100u since it's already extra storage when combined with the inserted beaker?

**Comment by CinnamonSnowball** [comment]
1000u is a bit crazy for storage, maybe tone it down to 300u (highest we typically go, with Bluespace Beakers and fermenting barrels), or even just 100u since it's already extra storage when combined with the inserted beaker?

**Comment by Migratingcocofruit** [comment]
> 1000u is a bit crazy for storage, maybe tone it down to 300u (highest we typically go, with Bluespace Beakers and fermenting barrels), or even just 100u since it's already extra storage when combined with the inserted beaker? Also, looking at the code more closely, you probably don't need a whole internal beaker for your tank purposes, you should just use a reagent datum

This is supposed to replace a chem dispesner, so starting at 1000 seems sensible. And I used a cryo beaker internally so you wouldn't have to worry about reactions in the machine, I'll try to implement it with a reagent datum and see how it works.

**Comment by Migratingcocofruit** [comment]
> Does it have a higher build cost or something? I can't find it quickly while glancing the code.

It's the same. Just gets more benefits from the matter bins and manipulators. I suppose adding some metal would make sense thematically, considering the way it looks.

**Comment by Migratingcocofruit** [comment]
> I don't think it's necessary to create a separate interface for this, you can just add a Tank section to ReagentGrinder if it's a BotaniTank. Look at smartfridge/dryer

The tank section is practically a copy of the smartfridge interface, so you're probably right. I'll try it and see how it goes.

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm:861]: These are all funcitons copied from the reagent grinder section and modified a bit for the Botanitank. I think it would be better to go over the entire file and I feel like this should be a second PR. There's also an infinite beaker glitch because the regular grinder constructor just makes one each time you build it, and funnily enough you have to remove the previous beaker to deconstruct it with a screw driver and crow bar. 
Either way I think I'll fix those separately.

**Comment by Migratingcocofruit** [comment]
> > You'd simply be left with less reagents this way, and you can always dump the excess. Also, this only occurs if you dispense all reagents proportionally(dispense mix button) which you wouldn't be doing without a specific purpose.
> > Perhaps I'll do something for the UI.
> 
> sort of my want, getting stuck with .564756 can get very frustrating with doing reactions

You can dump or dispense an arbitrary amount of any chem, so I'd rather leave that option with the player. It's not as though those fraction will use up too much of your tank, and if you made a particular mix I see no reason the machine will just lose parts of it each time you dispense it, not to mention it could ruin otherwise perfect proportions for reactions.

File: code/game/machinery/constructable_frame.dm
Comment (Migratingcocofruit) [reply code/game/machinery/constructable_frame.dm]: done

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: done

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: done

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm:857]: Good point. I'll just go over the entire file here then. no point only half fixing it.

File: code/game/machinery/constructable_frame.dm
Comment (Migratingcocofruit) [reply code/game/machinery/constructable_frame.dm:679]: I'll just go over the whole thing now.

**Comment by FunnyMan3595** [comment]
I'm against this until/unless we can find a way to make this not just a better version of the grinder with half a chemmaster and a huge tank to boot.

File: tgui/packages/tgui/interfaces/ReagentGrinder.js
Comment (AyIong) [inline tgui/packages/tgui/interfaces/ReagentGrinder.js:9]:
```suggestion
import { Box, Button, Section, Table, Stack, NumberInput, LabeledList, Icon } from '../components';
import { Window } from '../layouts';
import { BeakerContents } from '../interfaces/common/BeakerContents';
import { Operating } from '../interfaces/common/Operating';
```

File: tgui/packages/tgui/interfaces/ReagentGrinder.js
Comment (AyIong) [inline tgui/packages/tgui/interfaces/ReagentGrinder.js:164]:
```suggestion
          {!tank_contents && (
```
Use ternary 
{tank contents ? (Botanitank content) : (The Botanitank is empty content)}

**Comment by AyIong** [comment]
Also the UI is too big in my opinion, I would reduce it by switching between delete mode and move to container, like in ChemMaster
In this way you can reduce the width, well, and accordingly reduce the height, the Tank section should be scrollable in any case

File: tgui/packages/tgui/interfaces/ReagentGrinder.js
Comment (Migratingcocofruit) [reply tgui/packages/tgui/interfaces/ReagentGrinder.js:9]: done

File: tgui/packages/tgui/interfaces/ReagentGrinder.js
Comment (Migratingcocofruit) [reply tgui/packages/tgui/interfaces/ReagentGrinder.js:164]: done

**Comment by Migratingcocofruit** [comment]
> Also the UI is too big in my opinion, I would reduce it by switching between delete mode and move to container, like in ChemMaster In this way you can reduce the width, well, and accordingly reduce the height, the Tank section should be scrollable in any case

I shrunk the UI some. updated the photo with chem dispenser UI for scale.

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (AyIong) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:1]: 6 of what? Do please 0.6 SECONDS

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (AyIong) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:426]: I don't think spawn can be used

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:142]:
```suggestion
		icon_state = internal ? "botanitank_beaker" : "juicer1"
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:144]:
```suggestion
		icon_state = internal ? "botanitank_empty" : "juicer0"
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:162]:
```suggestion
	default_deconstruction_screwdriver(user, internal ? "botanitank_open" : "juicer_open", internal ? "botanitank_empty" : "juicer0", I)
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:410]:
```suggestion
	var/obj/item/reagent_containers/B
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:450]:
```suggestion
	var/obj/item/reagent_containers/B
```

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Oyu07) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm:19]:
```suggestion
	var/obj/item/reagent_containers/internal
```

**Review by Oyu07 — COMMENTED** [review]
These should already be null, no need to set them to null

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm:1]: 6 seconds, though it is the default value for a variable, so maybe PROCESS_TIME_SECONDS?

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm
Comment (Migratingcocofruit) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm:426]: > I don't think spawn can be used

spawn has always been used here, I only modified the time.

**Comment by Migratingcocofruit** [comment]
I've created [a branch](https://github.com/Migratingcocofruit/AddSoil/tree/By-your-powers-combined) that has the changes from this PR as well as https://github.com/ParadiseSS13/Paradise/pull/25050. Feel free to play test and give your input. 
My own impressions so far:
setup is about as quick as you can handle, though maturation time will still stall you when waiting for your first glowcaps and tower caps, as well as with the blumpkins.
Aside from that, if you plant enough plants to mutate with radium and heal them with the nutrient from the ground shrooms you can get all the plants you need fairly quickly and move onto mass producing your chems. setup is everything and it's easy to get overwhelmed so plan your actions and where you place your grinder and plants in advance so you don't have to rearrange things while also needing to take care of your plants. It's quite fun and engaging, aside from waiting for plants to mature, and rewards knowledge and planning, as it's quite easy to overwhelm yourself by taking care of plants you don't need or make a sub optimal setup.

File: code/game/machinery/constructable_frame.dm
Comment (Oyu07) [inline code/game/machinery/constructable_frame.dm:701]: Empty spaces,
698-700 delete
700-701 gap

Comment (Migratingcocofruit) [reply code/game/machinery/constructable_frame.dm:701]: > Edit: I think? You can use a switch instead of a for loop?

Wouldn't that be a problem if the grinder ever gets more children? Also what would be the advantage?

Comment (Oyu07) [reply code/game/machinery/constructable_frame.dm]: I mean, you can get rid of the for loop, and just use a switch, instead of doing 
`
for(var/name in ...)
  if(board_name == type)
    board_name = name

if(board_name == "botanik"
`

You can do
`switch(type)
    if("Botanik")
      ...
    if("all_in_one")
      ...
    board_name = name // if you need board_name
`

Advantage is optimization, and not sure if there is a need for for loop at all

Comment (Migratingcocofruit) [reply code/game/machinery/constructable_frame.dm]: I suppose if there are ever enough types of grinder that a loop is necessary it can be added then. very well.

**Review by Oyu07 — APPROVED** [review]

**Comment by Oyu07** [comment]
Ignore the 2 sorry*

**Review by S34NW — APPROVED** [review]

**Comment by S34NW** [comment]
Approving TGUI, code still needs reviewing more in-depth

**Comment by Stoniest** [comment]
Balance wise, there isn't a huge issue with it, but it does mean a lot of other machines become obsolete, there might be better ways to improve apon this. I'm staying neutral on this, if it isn't improved apon, then I'll be opposed to this. Maybe find ways to give it a bit more unique functionality for botany rather then a juiced up grinder that is also part chem master. It'll start making Chemistry in both medbay and science likely want to have something that is meant to be botany specific.

**Comment by Migratingcocofruit** [comment]
> Balance wise, there isn't a huge issue with it, but it does mean a lot of other machines become obsolete, there might be better ways to improve apon this. I'm staying neutral on this, if it isn't improved apon, then I'll be opposed to this. Maybe find ways to give it a bit more unique functionality for botany rather then a juiced up grinder that is also part chem master. It'll start making Chemistry in both medbay and science likely want to have something that is meant to be botany specific.
> 
> (Sorry for the late response)

I've been trying to think of a way to do so, but the main issue is the grinder is very basic, so any variant of it will likely make it obsolete, not to mention most of what it grinds are plants to begin with. As for the chem master, it's main purpose is making pills, not filtering chems, and that is done here in a way that is more similar to the chem dispenser anyhow so I wouldn't say the chem master becomes obsolete.   
The only thing that comes to mind is removing most sheets from it's grind list. leaving metal rods and planks as they can be grown.

**Comment by S34NW** [comment]
This currently has balance approval (+2) but not design (-2) or sprite (-1) teams. This is primarily a design PR so as of right now we're blocked from merge. I suggest you contact the design team to discuss changes.

Objection feedback anonymised:

- Basically makes two other machines useless. Not a great thing to make other mechanics trash.
- Until/unless we can find a way to make this not just a better version of the grinder with half a chemmaster and a huge tank to boot, I'm opposed.
- sprite is bad, i wont bother with it and help author improve it unless this moves into more of an approved stage

**Comment by Migratingcocofruit** [comment]
> This currently has balance approval (+2) but not design (-2) or sprite (-1) teams. This is primarily a design PR so as of right now we're blocked from merge. I suggest you contact the design team to discuss changes.
> 
> Objection feedback anonymised:
> 
>     * Basically makes two other machines useless. Not a great thing to make other mechanics trash.
> 
>     * Until/unless we can find a way to make this not just a better version of the grinder with half a chemmaster and a huge tank to boot, I'm opposed.
> 
>     * sprite is bad, i wont bother with it and help author improve it unless this moves into more of an approved stage

Any way I could contact the design team? I've made some changes and besides that I'd like to talk to them to see what can be done to improve it further if needed.

**Comment by Henri215** [comment]
> > This currently has balance approval (+2) but not design (-2) or sprite (-1) teams. This is primarily a design PR so as of right now we're blocked from merge. I suggest you contact the design team to discuss changes.
> > Objection feedback anonymised:
> > ```
> > * Basically makes two other machines useless. Not a great thing to make other mechanics trash.
> > 
> > * Until/unless we can find a way to make this not just a better version of the grinder with half a chemmaster and a huge tank to boot, I'm opposed.
> > 
> > * sprite is bad, i wont bother with it and help author improve it unless this moves into more of an approved stage
> > ```
> 
> Any way I could contact the design team? I've made some changes and besides that I'd like to talk to them to see what can be done to improve it further if needed.

Ping the design team over the discord, or ask for individuals to talk to.

**Comment by S34NW** [comment]
Unfortunately, the voting timeframe has closed for this PR and the design team are still opposed to this PR. Balance team are in favour.
In addition to the previous above comments, here are some additional ones shared:

- I actually really like this idea. Using the chem master with botany plants is a tedious pain and this would cure it. All of the functionality this provides is already given by a bunch of wooden barrels, a few beakers, and a chem master anyways, so I don't think there's any inherent balance concern here.
- As **REDACTED** said, balance wise, there isn't a huge issue with it, but it does mean a lot of other machines become obsolete, there might be better ways to improve apon this. I'm staying neutral on this, if it isn't improved apon, then I'll say no.

Despite balance approval, design refusal will block the merge of this PR as-is.
As such, I will be closing the PR.
Feel free to re-visit and make a new PR once you've had a chance to discuss with design and sort any issues.

**Pull request closed without merging**
